docs: add a "pack a Pro snap" how-to (#6367)
snapcore/snapcraft: docs/how-to/crafting/pack-a-pro-snap.rst

.. |app| replace:: Snapcraft
.. |app-command| replace:: snapcraft
.. |artifact| replace:: snap
.. |an-artifact| replace:: a snap
.. |app-min-pro-version| replace:: 9.1
.. |app-link| replace:: https://snapcraft.io/snapcraft

.. _how-to-pack-a-pro-snap:

.. include:: ../../common/craft-application/how-to-guides/pack-a-pro-artifact.rst

>     :end-before: .. Split point for downstream docs: Prerequisites onward

Reviewer: There's an inconsistency here. The title is transcluded, but the first heading isn't. If this is override is needed, then I don't think it's worth transcluding the title – may as well write it here, and transclude the two sections, without headings.

Author: https://github.com/canonical/snapcraft/pull/6367/commits/ffb2c985566bdfcb2dc2a4f1a082c7dca0e8585f

Nice catch, I simplified it to "begin overview" and "end overview". Hopefully that makes it easier to read the rst and is properly transcluded.

>     :end-before: .. Split point for downstream docs: Prerequisites onward

Reviewer: > onward

And upward?

Core24 and higher
~~~~~~~~~~~~~~~~~~~

.. include:: ../../common/craft-application/how-to-guides/pack-a-pro-artifact.rst
    :start-after: .. Split point for downstream docs: Prerequisites onward

Core22
~~~~~~~

Core22 snaps use a legacy mechanism to pack Pro-compliant snaps. As a result, it uses

> the previous name UA (Ubuntu Advantage) instead of Ubuntu Pro.

Reviewer: Forgive the nitpick that I didn't raise earlier:

```suggestion
Core22 snaps use a legacy mechanism to pack Pro-compliant snaps. The mechanism uses the 
term *Ubuntu Advantage* (UA) instead of *Ubuntu Pro*.
```

Author: https://github.com/canonical/snapcraft/pull/6367/commits/ffb2c985566bdfcb2dc2a4f1a082c7dca0e8585f

Prerequisites
-------------

- An Ubuntu Pro token (https://ubuntu.com/pro)

Identify the required Pro services
----------------------------------

First, determine which Pro services fit your needs:

- ``esm-apps`` or ``esm-infra``: If your goal is to pack a snap for an application
  and include the latest security patches for a base that is no longer under Standard
  Security Maintenance.
- ``fips``, ``fips-updates`` or ``fips-preview``: If you need to deploy a snap in a
  highly regulated environment that processes sensitive data.

The desired Pro services must be available. On a system with your Pro token
attached, run ``pro status`` and check the ``ENTITLED`` column for available
services. The Ubuntu Pro Client documentation has `further information
<https://documentation.ubuntu.com/pro-client/en/v32/explanations/which_services/>`__ on

> each service.

Reviewer: ```suggestion
services. The Ubuntu Pro Client documentation has `detailed information on each service
<https://documentation.ubuntu.com/pro-client/en/v32/explanations/which_services/>`__.
```

Reviewer: Alternatively, we could link this in the side bar with `seealso`.

Author: Interesting, I hadn't seen `seealso` before.  Like the `seealso`s on [this page](https://canonical.com/juju/docs/charmcraft/4.3/howto/manage-web-app-charms/configure-web-app-charm/)?  I don't see anything in the sidebar on that page for the links.

Author: https://github.com/canonical/snapcraft/pull/6367/commits/ffb2c985566bdfcb2dc2a4f1a082c7dca0e8585f (I went with the first suggestion in the interim)

> To enable the desired Pro services when building your snap, add a ua-services key to

Reviewer: ```suggestion
Add a ``ua-services`` key to your snap, and list the services:
```

Author: https://github.com/canonical/snapcraft/pull/6367/commits/ffb2c985566bdfcb2dc2a4f1a082c7dca0e8585f

.. |app| replace:: Snapcraft
.. |app-command| replace:: snapcraft
.. |artifact| replace:: snap
.. |an-artifact| replace:: a snap
.. |app-min-pro-version| replace:: 9.1
.. |app-link| replace:: https://snapcraft.io/snapcraft

.. _how-to-pack-a-pro-snap:

.. include:: ../../common/craft-application/how-to-guides/pack-a-pro-artifact.rst
    :end-before: .. Split point for downstream docs: Prerequisites onward

Core24 and higher
~~~~~~~~~~~~~~~~~~~

.. include:: ../../common/craft-application/how-to-guides/pack-a-pro-artifact.rst
    :start-after: .. Split point for downstream docs: Prerequisites onward

Core22
~~~~~~~

Core22 snaps use a legacy way to pack Pro-compliant snaps. It uses the term
"UA" (Ubuntu Advantage), the previous name for Ubuntu Pro.

> 

Author: Whoops, I forgot to apply the feedback from the scratchpad to the PR.

.. meta::
    :description: Pack a snap that contains extended security patches or meets regulatory compliance needs, powered by Ubuntu Pro.

.. |app| replace:: Snapcraft
.. |app-command| replace:: snapcraft
.. |artifact| replace:: snap
.. |an-artifact| replace:: a snap
.. |app-min-pro-version| replace:: 9.1
.. |app-link| replace:: https://snapcraft.io/snapcraft

.. _how-to-pack-a-pro-snap:

Pack a Pro-compliant snap
=========================

.. include:: ../../common/craft-application/how-to-guides/pack-a-pro-artifact.rst
    :start-after: .. Begin overview
    :end-before: .. End overview

Core24 and higher

> ~~~~~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
Core24 and higher
~~~~~~~~~~~~~~~~~
```

Author: https://github.com/canonical/snapcraft/pull/6367/commits/60a60f16fadc32108cc3cd12b0f3957cf3534976

.. include:: ../../common/craft-application/how-to-guides/pack-a-pro-artifact.rst
    :start-after: .. End overview

Core22
~~~~~~~

Core22 snaps use a legacy mechanism to pack Pro-compliant snaps. The mechanism uses the
term *Ubuntu Advantage* (UA) instead of *Ubuntu Pro*.

Prerequisites
-------------

- An Ubuntu Pro token (https://ubuntu.com/pro)

Identify the required Pro services
----------------------------------

First, determine which Pro services fit your needs:

- ``esm-apps`` or ``esm-infra``: If your goal is to pack a snap for an application
  and include the latest security patches for a base that is no longer under Standard
  Security Maintenance.
- ``fips``, ``fips-updates`` or ``fips-preview``: If you need to deploy a snap in a
  highly regulated environment that processes sensitive data.

The desired Pro services must be available. On a system with your Pro token
attached, run ``pro status`` and check the ``ENTITLED`` column for available
services. The Ubuntu Pro Client documentation has `detailed information on each service

> <https://documentation.ubuntu.com/pro-client/en/v32/explanations/which_services/>`__.

Reviewer 2: Looks like a delta was introduced between this paragraph and its transcluded counterpart.

Author: Good catch. This was reworded more recently, so I'll update craft-application to match it.

> Add a ``ua-services`` key to your snap, and list the services:

Reviewer 2: ```suggestion
List the desired services with the ``ua-services`` key in your project file:
```

Author: https://github.com/canonical/snapcraft/pull/6367/commits/60a60f16fadc32108cc3cd12b0f3957cf3534976

.. code-block:: yaml
    :caption: snapcraft.yaml

    ua-services:
      - esm-apps
      - esm-infra

Pack the snap
-------------

> Now you can pack the snap with the desired services by passing the Pro token:

Reviewer 2: ```suggestion
Pack your Pro-compliant snap with the ``pack`` command, providing the Pro token with the ``--ua-token`` option:
```
If you decide to implement https://github.com/canonical/craft-application/pull/1144/changes#r3754272113

Author: I sure did: https://github.com/canonical/snapcraft/pull/6367/commits/60a60f16fadc32108cc3cd12b0f3957cf3534976

.. code-block:: bash
    :substitutions:

    snapcraft pack --ua-token <pro-token> --enable-experimental-ua-services

The Pro token can also be provided as an environment variable:

.. code-block:: bash
    :substitutions:

    SNAPCRAFT_UA_TOKEN=<token> snapcraft pack --enable-experimental-ua-services

> As Pro enablement is an experimental feature, ``--enable-experimental-ua-services`` is

Reviewer 2: ```suggestion
As Pro enablement is an experimental feature, the ``--enable-experimental-ua-services`` option is
```

Author: https://github.com/canonical/snapcraft/pull/6367/commits/60a60f16fadc32108cc3cd12b0f3957cf3534976
